### PR TITLE
leave an empty scriptDirectory empty

### PR DESCRIPTION
**Pull request by acdimalev**
When loading an Emscripten binary (js + wasm) with jQuery, `document.currentScript.src` is an empty string, thus `scriptDirectory` is set to an empty string.

https://github.com/kripken/emscripten/blob/699d457f4ff55e0422ea79517714b07d691e8dae/src/shell.js#L234

However, the code for stripping the filename off the end of `scriptDirectory` translates the empty string into `/`.  This seems surprising / probably not preferred behavior.

https://github.com/kripken/emscripten/blob/699d457f4ff55e0422ea79517714b07d691e8dae/src/shell.js#L251

Minimum example for reproduction
https://gist.github.com/acdimalev/854b49353a439b7c5057fcff609a3aba

Since the code change to leave an empty `scriptDirectory` alone is almost more self-evident than this entire description and example, I'm starting with a pull request rather than a bug report.  If you'd like me to handle this any other way, please let me know.

### CODE REVIEW

**Comment by kripken** [comment]
Thanks @acdimalev!

This does look like something we need to fix, I agree. But I'm curious how `document.currentScript.src` could be empty - is that something jQuery is doing? (Regardless we should fix it.)

Notes on the PR:
 * Please add to the comment before that line, to mention why we check if the string is empty.
 * We should add a test for this. Maybe as part of, or alongside, `test_browser_run_from_different_directory` and the other last few tests in `tests/test_browser.py`, which all check the scriptDirectory/currentScript issue.
 * Please add yourself to AUTHORS.

**Comment by acdimalev** [comment]
So, it looks like this is exactly the case that `test_browser_modularize_no_current_script` checks for, except that the test is incorrectly passing because scriptDirectory should be empty, but is `/` instead, which just happens to work out because that's where the files actually are.

Working on a small modification to correct this test and will update this pull request once that's done.

File: src/shell.js [hunk 1]
```diff
@@ -247,7 +247,8 @@ if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
 #endif
   // blob urls look like blob:http://site.com/etc/etc and we cannot infer anything from them.
   // otherwise, slice off the final part of the url to find the script directory.
-  if (scriptDirectory.indexOf('blob:') !== 0) {
+  // be sure to leave an empty scriptDirectory alone.
+  if (scriptDirectory && scriptDirectory.indexOf('blob:') !== 0) {
```
Comment (juj) [inline src/shell.js]: How about

```javascript
if (scriptDirectory.indexOf('blob:') !== 0) {
  scriptDirectory = scriptDirectory.substr(0, scriptDirectory.lastIndexOf('/')+1);
} ...
```

which reads simpler overall than the original split+slice+join+'/' sequence? (and is 130 characters vs 143 chars, and less temp garbage generation)

Comment (acdimalev) [reply src/shell.js]: Logically it seems to make sense?  If `lastIndexOf` is guaranteed to return `-1` on all browsers when the needle is not found in the haystack, then it should result in taking a zero-length sub-string `(0, 0)`.  Otherwise it should result in the path leading up to the final slash.  Adding one to the returned index then works in both cases specifically because we want to include the slash in the resulting output.  For the case where a slash is present in the string (`scriptDirectory`), this would also require that both `lastIndexOf` and `substr` operate on the same index representation, be it bytes, code points, or otherwise.

These seem like logical assumptions, but this is the exact kind of program logic that I would personally want to write property tests to cross-validate my logical argument for why the code should work, especially given non-trivial inputs and a variety of javascript implementations.

If you think that's an important thing to change here, please let me know.  But before I dive in, would you mind giving me feedback on my proposed changes to the tests?

Thanks!

**Comment by acdimalev** [comment]
@kripken it's a bit wordy since filesystem and browser paths aren't necessarily the same thing, but just updated the test to run both variants.  This could be a bit shorter / simpler if you don't mind depending on Python to automatically convert slashes on Windows (e.g. `C:/example.txt`).  I'm not entirely sure what running both variants nets us for this test, but I don't mind erring on the side of caution.

**Comment by acdimalev** [comment]
@kripken & @juj, I wrote a few property tests for the shorter version of the code.

https://gist.github.com/acdimalev/e5f5a687879302e514d1f1b5fb4d8234

So far I've run the tests on Firefox `61.0`, Chromium `66.0.3359.170`, and Node JS `v9.10.0`.  Will run the tests on a few more browsers.  Assuming it checks out, I'll go ahead and update the code.

**Comment by kripken** [comment]
Test changes look good, thanks. (I agree that having both isn't obviously necessary, but yeah, extra testing for stuff like this is better for safety.)

**Comment by acdimalev** [comment]
@juj I'm reasonably convinced that your suggestion fully works as intended so I've gone ahead and updated accordingly.  I was able to run my property tests on Internet Explorer (though I'm not sure which version) and Edge `42.17134.1.0` with the aide of a couple polyfills by Mathias Bynens.

https://github.com/mathiasbynens/String.prototype.startsWith
https://github.com/mathiasbynens/String.prototype.includes
